hal/src/sercom/v2/uart/charsize.rs

```diff
@@ -0,0 +1,114 @@
+//! Character size definitions
+
+use crate::typelevel::Sealed;
+use num_traits::{AsPrimitive, PrimInt};
+
+/// Size of the SERCOM's `DATA` register
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+pub type DataSize = u16;
+
+/// Size of the SERCOM's `DATA` register
+#[cfg(any(feature = "min-samd51g"))]
+pub type DataSize = u32;
+
+/// Type-level `enum` representing the UART character size
+///
+/// The UART character size affects the word size for the embedded HAL traits.
+/// Eight or less bit transactions use a `u8` word, while nine-bit transactions
+/// use a `u16` word.
+pub trait CharSize: Sealed {
+    /// Word size for the character size
+    type Word: 'static + PrimInt + AsPrimitive<DataSize>;
+
+    /// Bits to write into the `LENGTH` register
+    const BITS: u8;
+}
+
+/// Type-level `enum` indicating a [`CharSize`] that is not dynamic
+pub trait FixedCharSize: CharSize {}
+
+/// Type alias to recover the `Word` type from an implementation of [`CharSize`]
+pub type Word<C> = <C as CharSize>::Word;
+
+/// [`CharSize`] variant for 5-bit transactions
+pub enum FiveBit {}
+
+/// [`CharSize`] variant for 6-bit transactions
+pub enum SixBit {}
+
+/// [`CharSize`] variant for 7-bit transactions
+pub enum SevenBit {}
+
+/// [`CharSize`] variant for 8-bit transactions
+pub enum EightBit {}
+
+/// [`CharSize`] variant for 9-bit transactions
+pub enum NineBit {}
+
+/// Dynamic [`CharSize`] that can be changed on the fly
+pub enum DynCharSize {}
+
+/// `enum` version of [`CharSize`]
+#[repr(u8)]
+pub enum CharSizeEnum {
+    FiveBit = FiveBit::BITS,
+    SixBit = SixBit::BITS,
+    SevenBit = SevenBit::BITS,
+    EightBit = EightBit::BITS,
+    NineBit = NineBit::BITS,
+}
+
+impl From<u8> for CharSizeEnum {
+    fn from(item: u8) -> CharSizeEnum {
+        match item {
+            FiveBit::BITS => CharSizeEnum::FiveBit,
+            SixBit::BITS => CharSizeEnum::SixBit,
+            SevenBit::BITS => CharSizeEnum::SevenBit,
+            EightBit::BITS => CharSizeEnum::EightBit,
+            NineBit::BITS => CharSizeEnum::NineBit,
+            _ => unreachable!(),
+        }
+    }
+}
```

Comment:
I don't know that you need this conversion. And if you do, I wouldn't implement it with `From`. I would use `TryFrom`. The `unreachable!` option is certainly not unreachable.

In the `spi` module, I realized that the only time I needed to convert from the `enum` to the raw bits is inside the corresponding `reg` function. So rather than abstract it into a trait that is only used once, I kept it isolated to those functions.

Reply:
Would you say panicking is acceptable behaviour if we encounter an invalid bit pattern? Normally this shouldn't happen as the driver guarantees a valid pattern should be in the register field.

Reply:
> Normally this shouldn't happen as the driver guarantees a valid pattern should be in the register field.

I guess that's my point. It sounds like you're saying that this code should only be used by the driver, because it can guarantee that the given bit pattern is valid before the conversion. But if that's the case, the conversion probably shouldn't be implemented as `From`, because that gives users access as well. A user could easily provide an invalid bit pattern and get to the `unreachable!` branch.

I was arguing that this code probably belongs in the `reg` module. The user-facing type is `CharSizeEnum`, and users shouldn't have to know or care about its corresponding hardware bit representation. And once it's in the `reg` module, you can truly guarantee that the branch is `unreachable!`.

Reply:
By "in the `reg`" module, I mean as part of one of the `Registers` methods.

Reply:
Okay, yeah, that makes total sense.